nighthawk/source/server/http_test_server_filter.cc

```diff
@@ -0,0 +1,83 @@
+#include <string>
+
+#include "http_test_server_filter.h"
+
+#include "absl/strings/numbers.h"
+
+#include "common/protobuf/utility.h"
+#include "envoy/server/filter_config.h"
+
+#include "nighthawk/api/server/response_options.pb.validate.h"
+
+namespace Nighthawk {
+namespace Server {
+
+HttpTestServerDecoderFilterConfig::HttpTestServerDecoderFilterConfig(
+    const nighthawk::server::ResponseOptions& proto_config)
+    : server_config_(proto_config) {}
+
+HttpTestServerDecoderFilter::HttpTestServerDecoderFilter(
+    HttpTestServerDecoderFilterConfigSharedPtr config)
+    : config_(config) {}
+
+HttpTestServerDecoderFilter::~HttpTestServerDecoderFilter() {}
+
+void HttpTestServerDecoderFilter::onDestroy() {}
+
+Envoy::Http::FilterHeadersStatus
+HttpTestServerDecoderFilter::decodeHeaders(Envoy::Http::HeaderMap& headers, bool) {
+  const auto* request_config_header = headers.get(TestServer::HeaderNames::get().TestServerConfig);
+  nighthawk::server::ResponseOptions base_config = config_->server_config();
+
+  std::string error_message = "";
+  if (request_config_header != nullptr) {
+    try {
+      nighthawk::server::ResponseOptions json_config;
+      Envoy::MessageUtil::loadFromJson(request_config_header->value().c_str(), json_config);
+      base_config.MergeFrom(json_config);
+      Envoy::MessageUtil::validate(base_config);
+    } catch (Envoy::EnvoyException exception) {
+      error_message = exception.what();
+    }
+  }
+
+  if (error_message.empty()) {
+    decoder_callbacks_->sendLocalReply(
+        static_cast<Envoy::Http::Code>(200), std::string(base_config.response_size(), 'a'),
+        [&base_config](Envoy::Http::HeaderMap& direct_response_headers) {
+          for (auto header_value_option : base_config.response_headers()) {
+            auto header = header_value_option.header();
+            auto lower_case_key = Envoy::Http::LowerCaseString(header.key());
+            if (!header_value_option.append().value()) {
+              direct_response_headers.remove(lower_case_key);
+            }
+            direct_response_headers.addCopy(lower_case_key, header.value());
+          }
+        },
+        absl::nullopt);
+  } else {
+    decoder_callbacks_->sendLocalReply(
+        static_cast<Envoy::Http::Code>(500),
+        fmt::format("test-server didn't understand the request: {}", error_message), nullptr,
```

Comment:
Is this tested?

Reply:
Yes, this code-path is hit in tests, checking we get a `500` response-code. Not checking the specific error messages though.

Reply:
Fair enough; it's probably also reasonable to verify the error as it's Envoy generated and should be relatively stable.